mlir/lib/Conversion/GPUToGENX/LowerGpuOpsToGENXOps.cpp

```diff
@@ -0,0 +1,177 @@
+      populateGpuRewritePatterns(patterns);
+      if (failed(applyPatternsAndFoldGreedily(m, std::move(patterns))))
+        return signalPassFailure();
+    }
+
+    LLVMTypeConverter converter(ctx, options);
+    populateGpuMemorySpaceAttributeConversions(
+        converter, [](gpu::AddressSpace space) {
+          switch (space) {
+          case gpu::AddressSpace::Global:
+            // TODO: use enumerations instead of hard coding the address space
+            // numbers:
+            //   return GENX::GENXMemorySpace::kGlobalMemorySpace
+            return 1;
+          case gpu::AddressSpace::Workgroup:
+            return 3;
+          case gpu::AddressSpace::Private:
+            return 5;
+          }
+          llvm_unreachable("unknown address space enum value");
+          return 0;
+        });
+
+    RewritePatternSet llvmPatterns(ctx);
+
+    mlir::arith::populateArithToLLVMConversionPatterns(converter, llvmPatterns);
+    populateVectorToLLVMConversionPatterns(converter, llvmPatterns);
+    cf::populateControlFlowToLLVMConversionPatterns(converter, llvmPatterns);
+    populateFuncToLLVMConversionPatterns(converter, llvmPatterns);
+    populateFinalizeMemRefToLLVMConversionPatterns(converter, llvmPatterns);
+    populateGpuToGENXConversionPatterns(converter, llvmPatterns);
+    LLVMConversionTarget target(getContext());
+    configureGpuToGENXConversionLegality(target);
+    if (failed(applyPartialConversion(m, target, std::move(llvmPatterns))))
+      signalPassFailure();
+  }
+};
+
+} // namespace
+
+void mlir::configureGpuToGENXConversionLegality(ConversionTarget &target) {
+  target.addIllegalOp<func::FuncOp>();
+  target.addLegalDialect<::mlir::LLVM::LLVMDialect>();
+  target.addLegalDialect<::mlir::GENX::GENXDialect>();
+  target.addIllegalDialect<gpu::GPUDialect>();
+  // TODO: Do we have equivalent operation in GENX ?
+  // target.addIllegalOp<LLVM::CosOp, LLVM::ExpOp, LLVM::Exp2Op, LLVM::FAbsOp,
+  //                      LLVM::FCeilOp, LLVM::FFloorOp, LLVM::LogOp,
+  //                      LLVM::Log10Op, LLVM::Log2Op, LLVM::PowOp, LLVM::SinOp,
+  //                      LLVM::SqrtOp>();
+
+  // TODO: Remove once we support replacing non-root ops.
+  target.addLegalOp<gpu::YieldOp, gpu::GPUModuleOp, gpu::ModuleEndOp>();
+}
+
+template <typename OpTy>
+static void populateOpPatterns(LLVMTypeConverter &converter,
+                               RewritePatternSet &patterns, StringRef f32Func,
+                               StringRef f64Func) {
+  patterns.add<ScalarizeVectorOpLowering<OpTy>>(converter);
+  patterns.add<OpToFuncCallLowering<OpTy>>(converter, f32Func, f64Func);
+}
+
+void mlir::populateGpuToGENXConversionPatterns(LLVMTypeConverter &converter,
+                                               RewritePatternSet &patterns) {
+  populateWithGenerated(patterns);
+  // TODO: Add patterns
+  //  patterns
+  //    .add<GPUIndexIntrinsicOpLowering<gpu::ThreadIdOp, GENX::ThreadIdXOp,
```

Comment:
The GPUIndexIntrinsicOpLowering patterns should already work with the current GENXOps.td.  Have you tried to enable them?